flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BufferWritingResultPartition.java

```diff
@@ -348,6 +348,7 @@ private void finishUnicastBufferBuilder(int targetSubpartition) {
         if (bufferBuilder != null) {
             numBytesOut.inc(bufferBuilder.finish());
             numBuffersOut.inc();
+            bufferBuilder.recycle();
```

Comment:
do we need both `finish()` and `recycle()`? Wouldn't single `close()` be enough?

Reply:
Yes, unfortunately, we need both because in case of calling `close()` on exception we want to recycle buffer but don't want to finish it. In fact, I see only one place where we definitely should distinguish these methods so maybe it will be possible to reduce it to one method but I am still not sure.

Reply:
Can you point to what is this place where it matters? From the top of my head it seems like it shouldn't be an issue.

Reply:
at least here org.apache.flink.runtime.checkpoint.channel.ChannelStateChunkReader#readChunk. 
And for example here org.apache.flink.runtime.checkpoint.channel.ResultSubpartitionRecoveredStateHandler#recover, right now we create BufferConsumer and only then finishing the BufferBuilder and it is possible to safely replace the finish by the close but I think more correctly in this method to finish BufferBuilder first, then create BufferConsumer and only then close BufferBuilder.

But in general, I believe it is possible to reduce these two methods to one (it is not a big problem even for readChunk because BufferConsumer is not created yet). But in this case, we will leave only one possible invariant - commit all changes only when BufferConsumer is created already. So filling the buffer then committing changes and only then creating BufferConsumer will be impossible.
So for example this code won't be work:
```
try(BufferBuilder bb = requestBufferBuilder()) {
   bb.append(..);
   read(bb.createBufferConsumer());
}
```
it will need to be rewritten like this:
```
BufferBuilder bb = requestBufferBuilder();
bb.append(..);
BufferConsumer bc = bb.createBufferConsumer();
bb.close();
read(bc);
```
In conclusion, I think that logically it is two different methods(different responsibilities), but for simplification, it is possible to have only one method if you think it makes sense.